File: internal/conformance/harness.go
// Package conformance is an RFC-driven HTTP conformance test suite for CalStakk.
//
// Every test boots the real server in-process via httptest.NewServer and fires
// raw HTTP requests, asserting exact wire-level behavior: status codes, response
// headers, and the XML bodies mandated by the relevant RFC.  Test names embed
// the RFC number and section so failures map directly to spec clauses.
//
// Red/green contract: tests fail for real until the backend implements the
// feature. `make conformance` shows the current pass/fail score. `make
// check-fast` (lint + build) stays green throughout build-out.
//
// Spec sources: specs/rfc*.txt (RFC text files in the repository root).
package conformance

import (
	"context"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/jpincas/calstakk/internal/backend"
	"github.com/jpincas/calstakk/internal/server"
	"github.com/jpincas/calstakk/internal/storage"
	"github.com/stretchr/testify/require"
)

// ─── XML namespaces ───────────────────────────────────────────────────────────

const (
	nsDAV    = "DAV:"
	nsCalDAV = "urn:ietf:params:xml:ns:caldav"
	nsCS     = "http://calendarserver.org/ns/"
)

// ─── CalDAV path constants ────────────────────────────────────────────────────

const (
	principalPath    = storage.PrincipalPath    // /calstakk
	calendarHomePath = storage.CalendarHomePath // /calstakk/calendars
)

func collectionPath(name string) string {
	return calendarHomePath + "/" + name
}

func objectPath(collection, uid string) string {
	return calendarHomePath + "/" + collection + "/" + uid + ".ics"
}

// ─── In-process test server ───────────────────────────────────────────────────

// srv wraps an in-process CalDAV server for conformance testing.
type srv struct {
	ts   *httptest.Server
	base string
}

// newServer boots an in-process CalDAV server backed by a temp directory.
// The server is closed automatically when the test ends.
func newServer(t *testing.T) *srv {
	t.Helper()
	store, err := storage.New(t.TempDir())
	require.NoError(t, err)
	be := backend.New(store)
	hs := server.New(be)
	ts := httptest.NewServer(hs)
	t.Cleanup(ts.Close)
	return &srv{ts: ts, base: ts.URL}
}

// ─── Raw HTTP helpers ─────────────────────────────────────────────────────────

// rawResp is the result of a raw HTTP round-trip.
type rawResp struct {
	StatusCode int
	Header     http.Header
	Body       string
}

// do issues an arbitrary HTTP method with optional headers and body.
// Redirects are followed (use doNoRedirect for redirect-checking tests).
func (s *srv) do(t *testing.T, method, path string, headers map[string]string, body string) *rawResp {
	t.Helper()
	return s.request(t, s.ts.Client(), method, path, headers, body)
}

// doNoRedirect issues a request without following redirects.
// The returned response reflects the raw 3xx redirect, not the final target.
func (s *srv) doNoRedirect(t *testing.T, method, path string, headers map[string]string, body string) *rawResp {
	t.Helper()
	client := &http.Client{
		CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
			return http.ErrUseLastResponse
		},
		Transport: s.ts.Client().Transport,
	}
	return s.request(t, client, method, path, headers, body)
}

func (s *srv) request(t *testing.T, client *http.Client, method, path string, headers map[string]string, body string) *rawResp {
	t.Helper()
	var bodyReader io.Reader
	if body != "" {
		bodyReader = strings.NewReader(body)
	}
	req, err := http.NewRequestWithContext(context.Background(), method, s.base+path, bodyReader)
	require.NoError(t, err)
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := client.Do(req)
	require.NoError(t, err)
	defer func() { _ = resp.Body.Close() }()
	data, err := io.ReadAll(resp.Body)
	require.NoError(t, err)
	return &rawResp{
		StatusCode: resp.StatusCode,
		Header:     resp.Header,
		Body:       string(data),
	}
}

// ─── Multistatus XML parsing ──────────────────────────────────────────────────

// Multistatus is a parsed DAV:multistatus response body.
type Multistatus struct {
	Responses []*MSResponse
}

// Response returns the DAV:response for the given href path, or nil.
func (ms *Multistatus) Response(href string) *MSResponse {
	for _, r := range ms.Responses {
		if r.Href == href {
			return r
		}
	}
	return nil
}

// Len returns the number of responses.
func (ms *Multistatus) Len() int { return len(ms.Responses) }

// MSResponse is a single DAV:response within a multistatus.
type MSResponse struct {
	Href  string
	props map[string]*MSProp // "{ns}local" → prop
}

// Prop returns the property with the given local name (searches all namespaces).
// Returns nil if the property is not present in any propstat.
func (r *MSResponse) Prop(local string) *MSProp {
	if r == nil {
		return nil
	}
	suffix := "}" + local
	for k, v := range r.props {
		if strings.HasSuffix(k, suffix) || k == local {
			return v
		}
	}
	return nil
}

// PropStatus returns the HTTP status code for the named property, or 0.
func (r *MSResponse) PropStatus(local string) int {
	p := r.Prop(local)
	if p == nil {
		return 0
	}
	return p.Status
}

// MSProp is a property entry from a DAV:propstat.
type MSProp struct {
	Status   int      // HTTP status from the enclosing propstat
	text     string   // concatenated character data
	children []string // local names of first-level child elements
}

// Text returns the plain-text content of the property value.
func (p *MSProp) Text() string {
	if p == nil {
		return ""
	}
	return p.text
}

// HasChild returns true if the property value contains a direct child element
// with the given local name.
func (p *MSProp) HasChild(local string) bool {
	if p == nil {
		return false
	}
	for _, c := range p.children {
		if c == local {
			return true
		}
	}
	return false
}

// ─── XML unmarshal intermediates ──────────────────────────────────────────────

type xmlMSInner struct {
	InnerXML string `xml:",innerxml"`
}

type xmlMSPropstat struct {
	Prop   xmlMSInner `xml:"prop"`
	Status string     `xml:"status"`
}

type xmlMSResponse struct {
	Href      string          `xml:"href"`
	Propstats []xmlMSPropstat `xml:"propstat"`
	Status    string          `xml:"status"`
}

type xmlMultistatus struct {
	XMLName   xml.Name        `xml:"multistatus"`
	Responses []xmlMSResponse `xml:"response"`
}

// parseMultistatus parses a DAV:multistatus XML body into a Multistatus.
// The test fails immediately if the XML is malformed.
func parseMultistatus(t *testing.T, body string) *Multistatus {
	t.Helper()
	var raw xmlMultistatus
	require.NoError(t, xml.Unmarshal([]byte(body), &raw),
		"parseMultistatus: invalid XML:\n%s", body)

	ms := &Multistatus{}
	for _, rr := range raw.Responses {
		mr := &MSResponse{
			Href:  rr.Href,
			props: make(map[string]*MSProp),
		}
		for _, ps := range rr.Propstats {
			code := parseHTTPStatus(ps.Status)
			for key, prop := range extractProps(ps.Prop.InnerXML, code) {
				mr.props[key] = prop
			}
		}
		ms.Responses = append(ms.Responses, mr)
	}
	return ms
}

// extractProps tokenises the inner XML of a <DAV:prop> element and returns a
// map of "{ns}local" → MSProp for each child property element.
// A namespace-declaration wrapper is prepended so prefix-qualified names
// resolve correctly even when outer xmlns declarations are absent.
func extractProps(innerXML string, statusCode int) map[string]*MSProp {
	props := make(map[string]*MSProp)
	if strings.TrimSpace(innerXML) == "" {
		return props
	}

	// Wrap with common namespace declarations so the decoder resolves prefixes
	// that were declared on ancestor elements in the original response.
	const nsWrap = `<X xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav" xmlns:CS="http://calendarserver.org/ns/" xmlns:A="http://apple.com/ns/ical/">`
	wrapped := nsWrap + innerXML + `</X>`

	dec := xml.NewDecoder(strings.NewReader(wrapped))
	depth := 0
	var curKey string
	var curText strings.Builder
	var curChildren []string

	for {
		tok, err := dec.Token()
		if err != nil {
			break
		}
		switch v := tok.(type) {
		case xml.StartElement:
			depth++
			switch depth {
			case 1:
				// The <X> wrapper — skip.
			case 2:
				// First-level child = a property element.
				curKey = xmlPropKey(v.Name)
				curText.Reset()
				curChildren = nil
			case 3:
				// Grandchild = direct child of the property value.
				curChildren = append(curChildren, v.Name.Local)
			}
		case xml.EndElement:
			switch {
			case depth == 2 && curKey != "":
				props[curKey] = &MSProp{
					Status:   statusCode,
					text:     strings.TrimSpace(curText.String()),
					children: curChildren,
				}
				curKey = ""
				curText.Reset()
				curChildren = nil
			}
			depth--
		case xml.CharData:
			if depth == 2 {
				curText.Write(v)
			}
		}
	}
	return props
}

// xmlPropKey returns a "{namespace}local" key for a property name.
func xmlPropKey(name xml.Name) string {
	if name.Space == "" {
		return name.Local
	}
	return "{" + name.Space + "}" + name.Local
}

// parseHTTPStatus extracts the numeric code from "HTTP/1.1 200 OK".
func parseHTTPStatus(status string) int {
	parts := strings.Fields(status)
	if len(parts) < 2 {
		return 0
	}
	code := 0
	for _, c := range parts[1] {
		if c < '0' || c > '9' {
			break
		}
		code = code*10 + int(c-'0')
	}
	return code
}

// ─── Header maps ──────────────────────────────────────────────────────────────

func depthHeader(d string) map[string]string { return map[string]string{"Depth": d} }
func xmlContentType() map[string]string {
	return map[string]string{"Content-Type": "application/xml; charset=utf-8"}
}
func calContentType() map[string]string {
	return map[string]string{"Content-Type": "text/calendar; charset=utf-8"}
}
func withHeaders(maps ...map[string]string) map[string]string {
	out := make(map[string]string)
	for _, m := range maps {
		for k, v := range m {
			out[k] = v
		}
	}
	return out
}

// ─── iCalendar fixture builders ───────────────────────────────────────────────

const (
	testDTSTAMP = "20260101T000000Z"
	testDTSTART = "20260115T100000Z"
	testDTEND   = "20260115T110000Z"
	testDUE     = "20260120T120000Z"
)

// vtodo returns a minimal valid VTODO VCALENDAR string.
// Extra iCalendar property lines (e.g. "STATUS:COMPLETED\r\n") may be appended.
func vtodo(uid string, extra ...string) string {
	var sb strings.Builder
	sb.WriteString("BEGIN:VCALENDAR\r\n")
	sb.WriteString("VERSION:2.0\r\n")
	sb.WriteString("PRODID:-//CalStakk//ConformanceTest//EN\r\n")
	sb.WriteString("BEGIN:VTODO\r\n")
	fmt.Fprintf(&sb, "UID:%s\r\n", uid)
	fmt.Fprintf(&sb, "DTSTAMP:%s\r\n", testDTSTAMP)
	sb.WriteString("SUMMARY:Test Todo\r\n")
	for _, line := range extra {
		line = strings.TrimRight(line, "\r\n")
		sb.WriteString(line + "\r\n")
	}
	sb.WriteString("END:VTODO\r\n")
	sb.WriteString("END:VCALENDAR\r\n")
	return sb.String()
}

// vevent returns a minimal valid VEVENT VCALENDAR string.
func vevent(uid string, extra ...string) string {
	var sb strings.Builder
	sb.WriteString("BEGIN:VCALENDAR\r\n")
	sb.WriteString("VERSION:2.0\r\n")
	sb.WriteString("PRODID:-//CalStakk//ConformanceTest//EN\r\n")
	sb.WriteString("BEGIN:VEVENT\r\n")
	fmt.Fprintf(&sb, "UID:%s\r\n", uid)
	fmt.Fprintf(&sb, "DTSTAMP:%s\r\n", testDTSTAMP)
	fmt.Fprintf(&sb, "DTSTART:%s\r\n", testDTSTART)
	fmt.Fprintf(&sb, "DTEND:%s\r\n", testDTEND)
	sb.WriteString("SUMMARY:Test Event\r\n")
	for _, line := range extra {
		line = strings.TrimRight(line, "\r\n")
		sb.WriteString(line + "\r\n")
	}
	sb.WriteString("END:VEVENT\r\n")
	sb.WriteString("END:VCALENDAR\r\n")
	return sb.String()
}

// ─── PROPFIND / REPORT XML body builders ──────────────────────────────────────

func propfindAllprop() string {
	return xml.Header + `<D:propfind xmlns:D="DAV:"><D:allprop/></D:propfind>`
}

func propfindPropname() string {
	return xml.Header + `<D:propfind xmlns:D="DAV:"><D:propname/></D:propfind>`
}

// propfindProps builds a PROPFIND body requesting specific properties.
// Provide pairs of (namespace-URI, local-name); DAV: and CalDAV namespaces
// are automatically prefixed as D: and C: respectively.
func propfindProps(pairs ...string) string {
	var sb strings.Builder
	sb.WriteString(xml.Header)
	sb.WriteString(`<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav"><D:prop>`)
	for i := 0; i+1 < len(pairs); i += 2 {
		ns, local := pairs[i], pairs[i+1]
		switch ns {
		case nsDAV:
			fmt.Fprintf(&sb, `<D:%s/>`, local)
		case nsCalDAV:
			fmt.Fprintf(&sb, `<C:%s/>`, local)
		default:
			fmt.Fprintf(&sb, `<%s xmlns="%s"/>`, local, ns)
		}
	}
	sb.WriteString(`</D:prop></D:propfind>`)
	return sb.String()
}

// calendarQueryByComp builds a calendar-query REPORT body for a specific
// component type (e.g. "VTODO", "VEVENT").
func calendarQueryByComp(compName string) string {
	return xml.Header + fmt.Sprintf(
		`<C:calendar-query xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">`+
			`<D:prop><D:getetag/><C:calendar-data/></D:prop>`+
			`<C:filter><C:comp-filter name="VCALENDAR">`+
			`<C:comp-filter name="%s"/>`+
			`</C:comp-filter></C:filter>`+
			`</C:calendar-query>`,
		compName,
	)
}

// calendarQueryByTimeRange builds a calendar-query REPORT with a time-range filter.
func calendarQueryByTimeRange(compName, start, end string) string {
	return xml.Header + fmt.Sprintf(
		`<C:calendar-query xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">`+
			`<D:prop><D:getetag/><C:calendar-data/></D:prop>`+
			`<C:filter><C:comp-filter name="VCALENDAR">`+
			`<C:comp-filter name="%s">`+
			`<C:time-range start="%s" end="%s"/>`+
			`</C:comp-filter>`+
			`</C:comp-filter></C:filter>`+
			`</C:calendar-query>`,
		compName, start, end,
	)
}

// calendarMultiget builds a calendar-multiget REPORT body for the given hrefs.
func calendarMultiget(hrefs ...string) string {
	var sb strings.Builder
	sb.WriteString(xml.Header)
	sb.WriteString(`<C:calendar-multiget xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">`)
	sb.WriteString(`<D:prop><D:getetag/><C:calendar-data/></D:prop>`)
	for _, h := range hrefs {
		fmt.Fprintf(&sb, `<D:href>%s</D:href>`, h)
	}
	sb.WriteString(`</C:calendar-multiget>`)
	return sb.String()
}

// syncCollectionReport builds a sync-collection REPORT body.
// Pass an empty syncToken for the initial (full) sync.
func syncCollectionReport(syncToken string) string {
	return xml.Header + fmt.Sprintf(
		`<D:sync-collection xmlns:D="DAV:"><D:sync-token>%s</D:sync-token>`+
			`<D:sync-level>1</D:sync-level>`+
			`<D:prop><D:getetag/></D:prop></D:sync-collection>`,
		syncToken,
	)
}

// ─── Setup helpers ────────────────────────────────────────────────────────────

// mkcol creates a calendar collection using MKCOL and asserts 201 Created.
// Returns the collection path.
func (s *srv) mkcol(t *testing.T, name string) string {
	t.Helper()
	path := collectionPath(name)
	resp := s.do(t, "MKCOL", path, nil, "")
	require.Equal(t, http.StatusCreated, resp.StatusCode,
		"MKCOL %s failed: %s", path, resp.Body)
	return path
}

// putObject PUTs a calendar object at the given path. Returns the ETag.
func (s *srv) putObject(t *testing.T, path, body string) string {
	t.Helper()
	resp := s.do(t, "PUT", path, calContentType(), body)
	require.Equal(t, http.StatusCreated, resp.StatusCode,
		"PUT %s failed (status %d): %s", path, resp.StatusCode, resp.Body)
	return resp.Header.Get("ETag")
}

// putTodo creates a VTODO at objectPath(collection, uid) and returns the ETag.
func (s *srv) putTodo(t *testing.T, collection, uid string, extra ...string) string {
	t.Helper()
	return s.putObject(t, objectPath(collection, uid), vtodo(uid, extra...))
}

// putEvent creates a VEVENT at objectPath(collection, uid) and returns the ETag.
func (s *srv) putEvent(t *testing.T, collection, uid string, extra ...string) string {
	t.Helper()
	return s.putObject(t, objectPath(collection, uid), vevent(uid, extra...))
}

File: tests/conformance/harness.ts
// Conformance test harness — mirrors internal/conformance/harness.go.
//
// Each RFC test file boots a real in-process CalDAV server via Deno.serve,
// fires raw HTTP requests, and asserts wire-level behavior (status codes,
// response headers, XML bodies).

import { assertEquals } from "@std/assert";
import { createHandler } from "../../src/protocol.ts";
import { MemoryStorage } from "../../src/storage.ts";
import { hashPassword } from "../../src/auth.ts";
import { DEFAULT_CALENDAR_NAME } from "../../src/types.ts";
import type { Config } from "../../src/config.ts";
import { find, findAll, findDeep, parseXML } from "../../src/xmlparse.ts";

const NS_DAV = "DAV:";

// ─── Test owner credentials ───────────────────────────────────────────────────

// Owner acts as organizer in scheduling tests; email matches ORGANIZER: in ICS fixtures.
const OWNER_USERNAME = "owner";
const OWNER_PASSWORD = "ownerpass";
export const OWNER_EMAIL = "org@example.com";

// ─── CalDAV path constants ────────────────────────────────────────────────────

export const principalPath = `/principals/${OWNER_USERNAME}`;
export const calendarHomePath = `/calendars/${OWNER_USERNAME}`;

export function collectionPath(name: string): string {
  return `${calendarHomePath}/${name}`;
}

export function objectPath(collection: string, uid: string): string {
  return `${calendarHomePath}/${collection}/${uid}.ics`;
}

/** Build a calendar path for a non-owner user. */
export function userObjectPath(username: string, collection: string, uid: string): string {
  return `/calendars/${username}/${collection}/${uid}.ics`;
}

export function userCollectionPath(username: string, collection: string): string {
  return `/calendars/${username}/${collection}`;
}

export function userPrincipalPath(username: string): string {
  return `/principals/${username}`;
}

export function userInboxPath(username: string): string {
  return `/calendars/${username}/inbox`;
}

// ─── iCalendar fixture builders ───────────────────────────────────────────────

export const testDTSTAMP = "20260101T000000Z";
export const testDTSTART = "20260115T100000Z";
export const testDTEND = "20260115T110000Z";
export const testDUE = "20260120T120000Z";

/** Returns a minimal valid VTODO VCALENDAR string. */
export function vtodo(uid: string, ...extra: string[]): string {
  let s = "BEGIN:VCALENDAR\r\n";
  s += "VERSION:2.0\r\n";
  s += "PRODID:-//CalStakk//ConformanceTest//EN\r\n";
  s += "BEGIN:VTODO\r\n";
  s += `UID:${uid}\r\n`;
  s += `DTSTAMP:${testDTSTAMP}\r\n`;
  s += "SUMMARY:Test Todo\r\n";
  for (const line of extra) {
    s += line.replace(/\r?\n$/, "") + "\r\n";
  }
  s += "END:VTODO\r\n";
  s += "END:VCALENDAR\r\n";
  return s;
}

/** Returns a minimal valid VEVENT VCALENDAR string. Extra lines override defaults when they share the same property name. */
export function vevent(uid: string, ...extra: string[]): string {
  const hasOverride = (key: string) => extra.some((l) => l.toUpperCase().startsWith(`${key}:`));
  let s = "BEGIN:VCALENDAR\r\n";
  s += "VERSION:2.0\r\n";
  s += "PRODID:-//CalStakk//ConformanceTest//EN\r\n";
  s += "BEGIN:VEVENT\r\n";
  s += `UID:${uid}\r\n`;
  s += `DTSTAMP:${testDTSTAMP}\r\n`;
  if (!hasOverride("DTSTART")) s += `DTSTART:${testDTSTART}\r\n`;
  if (!hasOverride("DTEND")) s += `DTEND:${testDTEND}\r\n`;
  s += "SUMMARY:Test Event\r\n";
  for (const line of extra) {
    s += line.replace(/\r?\n$/, "") + "\r\n";
  }
  s += "END:VEVENT\r\n";
  s += "END:VCALENDAR\r\n";
  return s;
}

// ─── PROPFIND / REPORT XML body builders ──────────────────────────────────────

const XML_HEADER = '<?xml version="1.0" encoding="UTF-8"?>\n';

export function propfindAllprop(): string {
  return XML_HEADER + '<D:propfind xmlns:D="DAV:"><D:allprop/></D:propfind>';
}

export function propfindPropname(): string {
  return XML_HEADER + '<D:propfind xmlns:D="DAV:"><D:propname/></D:propfind>';
}

export const nsDAV = "DAV:";
export const nsCalDAV = "urn:ietf:params:xml:ns:caldav";

/**
 * Build a PROPFIND body for specific properties.
 * Provide pairs of [namespace, local-name].
 */
export function propfindProps(...pairs: string[]): string {
  let inner = "";
  for (let i = 0; i + 1 < pairs.length; i += 2) {
    const [ns, local] = [pairs[i], pairs[i + 1]];
    if (ns === nsDAV) inner += `<D:${local}/>`;
    else if (ns === nsCalDAV) inner += `<C:${local}/>`;
    else inner += `<X:${local} xmlns:X="${ns}"/>`;
  }
  return (
    XML_HEADER +
    `<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav"><D:prop>${inner}</D:prop></D:propfind>`
  );
}

export function calendarQueryByComp(compName: string): string {
  return (
    XML_HEADER +
    `<C:calendar-query xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">` +
    `<D:prop><D:getetag/><C:calendar-data/></D:prop>` +
    `<C:filter><C:comp-filter name="VCALENDAR">` +
    `<C:comp-filter name="${compName}"/>` +
    `</C:comp-filter></C:filter>` +
    `</C:calendar-query>`
  );
}

export function calendarQueryByTimeRange(
  compName: string,
  start: string,
  end: string,
): string {
  return (
    XML_HEADER +
    `<C:calendar-query xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">` +
    `<D:prop><D:getetag/><C:calendar-data/></D:prop>` +
    `<C:filter><C:comp-filter name="VCALENDAR">` +
    `<C:comp-filter name="${compName}">` +
    `<C:time-range start="${start}" end="${end}"/>` +
    `</C:comp-filter>` +
    `</C:comp-filter></C:filter>` +
    `</C:calendar-query>`
  );
}

export function calendarMultiget(...hrefs: string[]): string {
  const hrefXML = hrefs.map((h) => `<D:href>${h}</D:href>`).join("");
  return (
    XML_HEADER +
    `<C:calendar-multiget xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">` +
    `<D:prop><D:getetag/><C:calendar-data/></D:prop>` +
    hrefXML +
    `</C:calendar-multiget>`
  );
}

export function syncCollectionReport(syncToken: string): string {
  return (
    XML_HEADER +
    `<D:sync-collection xmlns:D="DAV:">` +
    `<D:sync-token>${syncToken}</D:sync-token>` +
    `<D:sync-level>1</D:sync-level>` +
    `<D:prop><D:getetag/></D:prop>` +
    `</D:sync-collection>`
  );
}

// ─── Header helpers ───────────────────────────────────────────────────────────

export function depthHeader(d: string): Record<string, string> {
  return { Depth: d };
}

export function xmlContentType(): Record<string, string> {
  return { "Content-Type": "application/xml; charset=utf-8" };
}

export function calContentType(): Record<string, string> {
  return { "Content-Type": "text/calendar; charset=utf-8" };
}

export function withHeaders(...maps: Record<string, string>[]): Record<string, string> {
  return Object.assign({}, ...maps);
}

// ─── Multistatus XML parsing ──────────────────────────────────────────────────

import type { XNode } from "../../src/xmlparse.ts";

/** Collect all text from a node and its descendants (depth-first). */
function allText(node: XNode): string {
  let t = node.text;
  for (const c of node.children) t += allText(c);
  return t;
}

export class MSProp {
  constructor(
    public readonly status: number,
    private readonly _text: string,
    private readonly _children: string[],
  ) {}

  text(): string {
    return this._text;
  }

  hasChild(local: string): boolean {
    return this._children.includes(local);
  }
}

export class MSResponse {
  constructor(
    public readonly href: string,
    private readonly props: Map<string, MSProp>,
  ) {}

  prop(local: string): MSProp | undefined {
    // Search by local name across all namespaces
    for (const [key, val] of this.props) {
      if (key === local || key.endsWith(`:${local}`) || key.endsWith(`}${local}`)) {
        return val;
      }
    }
    return undefined;
  }

  propStatus(local: string): number {
    return this.prop(local)?.status ?? 0;
  }
}

export class Multistatus {
  constructor(private readonly responses: MSResponse[]) {}

  response(href: string): MSResponse | undefined {
    return this.responses.find((r) => r.href === href);
  }

  len(): number {
    return this.responses.length;
  }
}

export function parseMultistatus(body: string): Multistatus {
  const root = parseXML(body);
  if (!root) return new Multistatus([]);

  const responses: MSResponse[] = [];
  const responseEls = findAll(root, NS_DAV, "response");

  for (const responseEl of responseEls) {
    const hrefEl = find(responseEl, NS_DAV, "href");
    const href = hrefEl?.text.trim() ?? "";

    const props = new Map<string, MSProp>();
    const propstatEls = findAll(responseEl, NS_DAV, "propstat");

    for (const propstatEl of propstatEls) {
      const statusEl = find(propstatEl, NS_DAV, "status");
      const code = parseStatusCode(statusEl?.text.trim() ?? "");

      const propEl = find(propstatEl, NS_DAV, "prop");
      if (!propEl) continue;

      for (const child of propEl.children) {
        const childNames = child.children.map((c) => c.local);
        const msProp = new MSProp(code, allText(child).trim(), childNames);
        props.set(`${child.ns}:${child.local}`, msProp);
        props.set(child.local, msProp); // index by local name too
      }
    }

    // Response-level status (used for 404 in multiget with no propstat)
    if (propstatEls.length === 0) {
      const statusEl = find(responseEl, NS_DAV, "status");
      if (statusEl) {
        const code = parseStatusCode(statusEl.text.trim());
        props.set("__status__", new MSProp(code, "", []));
      }
    }

    responses.push(new MSResponse(href, props));
  }

  return new Multistatus(responses);
}

/** Extract the sync-token value from a multistatus XML body. */
export function extractSyncToken(body: string): string {
  const root = parseXML(body);
  if (!root) return "";
  return findDeep(root, NS_DAV, "sync-token")?.text.trim() ?? "";
}

function parseStatusCode(statusLine: string): number {
  const parts = statusLine.split(" ");
  if (parts.length < 2) return 0;
  return parseInt(parts[1]) || 0;
}

// ─── Test server ──────────────────────────────────────────────────────────────

interface RawResponse {
  status: number;
  headers: Headers;
  body: string;
}

export class TestServer {
  private readonly baseURL: string;
  private readonly server: Deno.HttpServer<Deno.NetAddr>;
  private readonly controller: AbortController;
  private readonly storage: MemoryStorage;

  private constructor(
    baseURL: string,
    server: Deno.HttpServer<Deno.NetAddr>,
    controller: AbortController,
    storage: MemoryStorage,
  ) {
    this.baseURL = baseURL;
    this.server = server;
    this.controller = controller;
    this.storage = storage;
  }

  static async create(): Promise<TestServer> {
    const storage = new MemoryStorage();

    // Pre-create owner user with hashed password and default calendar.
    const ownerHash = await hashPassword(OWNER_PASSWORD);
    await storage.createUser({
      username: OWNER_USERNAME,
      passwordHash: ownerHash,
      displayName: "Test Owner",
      email: OWNER_EMAIL,
      timezone: "UTC",
      isAdmin: true,
    });
    await storage.createCalendar(OWNER_USERNAME, DEFAULT_CALENDAR_NAME, "Default Calendar");

    const testConfig: Config = {
      server: { host: "localhost", port: 0, kvPath: undefined, webDir: undefined },
      user: {
        username: OWNER_USERNAME,
        password: OWNER_PASSWORD,
        displayName: "Test Owner",
        email: OWNER_EMAIL,
        timezone: "UTC",
      },
    };

    const handler = createHandler(storage, testConfig);
    const controller = new AbortController();
    const server = Deno.serve(
      {
        port: 0,
        signal: controller.signal,
        onListen: () => {}, // suppress startup log
      },
      handler,
    );
    const port = server.addr.port;
    return new TestServer(`http://localhost:${port}`, server, controller, storage);
  }

  async shutdown(): Promise<void> {
    this.controller.abort();
    await this.server.finished;
  }

  /** Create a secondary user (non-owner) in the server's storage. */
  async createUser(username: string, email: string, password = "testpass"): Promise<void> {
    const hash = await hashPassword(password);
    await this.storage.createUser({
      username,
      passwordHash: hash,
      displayName: username,
      email,
      timezone: "UTC",
      isAdmin: false,
    });
    await this.storage.createCalendar(username, DEFAULT_CALENDAR_NAME, "Default Calendar");
  }

  /** Issue an HTTP request as the owner. Redirects are followed. */
  async do(
    method: string,
    path: string,
    headers: Record<string, string> = {},
    body?: string,
  ): Promise<RawResponse> {
    const ownerAuth = "Basic " + btoa(`${OWNER_USERNAME}:${OWNER_PASSWORD}`);
    const resp = await fetch(`${this.baseURL}${path}`, {
      method,
      headers: { Authorization: ownerAuth, ...headers },
      body: body ?? undefined,
      redirect: "follow",
    });
    return { status: resp.status, headers: resp.headers, body: await resp.text() };
  }

  /** Issue an HTTP request as a specific user. Redirects are followed. */
  async doAs(
    username: string,
    password: string,
    method: string,
    path: string,
    headers: Record<string, string> = {},
    body?: string,
  ): Promise<RawResponse> {
    const authHeader = "Basic " + btoa(`${username}:${password}`);
    const resp = await fetch(`${this.baseURL}${path}`, {
      method,
      headers: { Authorization: authHeader, ...headers },
      body: body ?? undefined,
      redirect: "follow",
    });
    return { status: resp.status, headers: resp.headers, body: await resp.text() };
  }

  /** Issue an HTTP request WITHOUT following redirects and WITHOUT auth (for 401 tests). */
  async doNoRedirect(
    method: string,
    path: string,
    headers: Record<string, string> = {},
    body?: string,
  ): Promise<RawResponse> {
    const resp = await fetch(`${this.baseURL}${path}`, {
      method,
      headers,
      body: body ?? undefined,
      redirect: "manual",
    });
    return { status: resp.status, headers: resp.headers, body: await resp.text() };
  }

  /** Create a calendar collection (MKCOL) for owner and assert 201. */
  async mkcol(name: string): Promise<string> {
    const path = collectionPath(name);
    const resp = await this.do("MKCOL", path);
    assertEquals(resp.status, 201, `MKCOL ${path} failed: ${resp.body}`);
    return path;
  }

  /** Create a MKCALENDAR collection for a specific user. */
  async mkcalAs(username: string, password: string, colName: string): Promise<string> {
    const path = userCollectionPath(username, colName);
    const resp = await this.doAs(username, password, "MKCALENDAR", path);
    assertEquals(resp.status, 201, `MKCALENDAR ${path} failed: ${resp.body}`);
    return path;
  }

  /** PUT a calendar object and assert 201. Returns ETag. */
  async putObject(path: string, body: string): Promise<string> {
    const resp = await this.do(
      "PUT",
      path,
      { "Content-Type": "text/calendar; charset=utf-8" },
      body,
    );
    assertEquals(
      resp.status,
      201,
      `PUT ${path} failed (${resp.status}): ${resp.body}`,
    );
    return resp.headers.get("ETag") ?? "";
  }

  /** PUT a VTODO and return the ETag. */
  putTodo(collection: string, uid: string, ...extra: string[]): Promise<string> {
    return this.putObject(objectPath(collection, uid), vtodo(uid, ...extra));
  }

  /** PUT a VEVENT and return the ETag. */
  putEvent(collection: string, uid: string, ...extra: string[]): Promise<string> {
    return this.putObject(objectPath(collection, uid), vevent(uid, ...extra));
  }

  get base(): string {
    return this.baseURL;
  }
}

/** Convenience wrapper: create a server, run fn, then shut down. */
export async function withServer(
  fn: (s: TestServer) => Promise<void>,
): Promise<void> {
  const s = await TestServer.create();
  try {
    await fn(s);
  } finally {
    await s.shutdown();
  }
}
